src/loop/parser.ml

```diff
@@ -119,12 +125,16 @@ module Make(State : State.S) = struct
       | Some `Incremental -> State.warn ?loc st full_mode_switch lang
       | _ -> st
     in
-    State.set State.logic_file { file with mode = Some `Full; } st
+    st, { file with mode = Some `Full }
 
-  let set_logic_file ?loc st old (file : _ file) =
+  let switch_to_full_mode_if_needed ?loc st old (file : _ file) =
     match file.lang with
     | Some Logic.Alt_ergo -> switch_to_full_mode ?loc "Alt-Ergo" st old file
-    | _ -> State.set State.logic_file file st
+    | _ -> st, file
+
+  let set_logic_file ?loc st old (file : _ file) =
+    let st, file = switch_to_full_mode_if_needed ?loc st old file in
+    State.set State.logic_file file st
```

Comment:
I was confused for a moment because I thought you referred to `switch_to_full_mode_if_needed` but yeah, that's correct — I will make it a local function.